### PR TITLE
Populate initial value from dropdown list for questionnaire items with type reference

### LINKED ISSUE #1830: Use initial expressions to select the initial value from a dropdown list populated using answer expressions
**Is your feature request related to a problem? Please describe.**
When I populate a dropdown using the answer expression extension as per this [example](https://github.com/google/android-fhir/pull/1480#issue-1287687814), I cannot set the initial value using the initial expression extension. This is due to the item being of a reference type and the type I'm trying to populate is a string type. 

**Describe the solution you'd like**
I should be able to use initial expressions to set the initial value

**Describe alternatives you've considered**
Instead of using answer expressions, I can populate the dropdown with answer options instead but this is challenging where there are a large number of options.

**Additional context**
Here is the [questionnaire](https://github.com/google/android-fhir/files/10508042/initial.expressions.plus.answer.expressions.sample.questionnaire.txt) I tried and this is the [logcat](https://github.com/google/android-fhir/files/10508046/initial.expressions.plus.answer.expressions.log.cat.txt). If there is another way to do this, please point me in the right direction.

### PR DESCRIPTION
**Pull request by LZRS**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1830, and close #1831 

**Description**
Adds support for setting initial values using initial expressions for questionnaire field items of type `reference` in the following forms

_Using Resource.id expression_
```
   },
        {
          "url": "http://hl7.org/fhir/uv/sdc/StructureDefinition/sdc-questionnaire-initialExpression",
          "valueExpression": {
            "language": "text/fhirpath",
            "expression": "Patient.id"
          }
        }
```

and _Using Resource_
```,
       {
          "url": "http://hl7.org/fhir/uv/sdc/StructureDefinition/sdc-questionnaire-initialExpression",
          "valueExpression": {
            "language": "text/fhirpath",
            "expression": "Patient"
          }
        },
```

Also adds fix for the autocomplete textview being set to empty when a dropdown item (using the `choiceColumn`) is selected. Seems to have been occuring when the `displayString` no longer evaluated to a unique string and `.singleOrNull()` always evaluated to null

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: (Bug fix | Feature | Documentation | Testing | Code health | Builds | Releases | Other)

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreTypes.kt [hunk 1]
```diff
@@ -94,6 +94,18 @@ fun Type.displayString(context: Context): String =
     else -> context.getString(R.string.not_answered)
   }
 
+/**
+ * Returns the unique identifier of a [Type]. Used to differentiate between item answer options that
+ * may have similar display strings
+ */
+fun Type.identifierString(context: Context): String =
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreTypes.kt:101]: do we need this function? check what I've done in the DropDownViewHolderFactory on how to get rid of it

**Comment by omarismail94** [comment]
@LZRS  Can you add a description of what this change does in the Description section above?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DropDownViewHolderFactory.kt [hunk 2]
```diff
@@ -167,6 +177,7 @@ internal class AnswerOptionDropDownArrayAdapter(
 }
 
 internal data class DropDownAnswerOption(
+  val answerIdentifierString: String,
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DropDownViewHolderFactory.kt:180]:
```suggestion
  val answerId: String,
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DropDownViewHolderFactory.kt [hunk 3]
```diff
@@ -67,19 +68,26 @@ internal object DropDownViewHolderFactory :
           this.questionnaireViewItem.answerOption
             .map {
               DropDownAnswerOption(
+                it.value.identifierString(context),
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DropDownViewHolderFactory.kt:71]: what are the cases when Type.id is null? Can you list a couple of examples? If it makes sense, then we can do the change below:

```suggestion
                it.value.let { it.id ?: it.displayString(context)},
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DropDownViewHolderFactory.kt [hunk 4]
```diff
@@ -67,19 +68,26 @@ internal object DropDownViewHolderFactory :
...
+            context.getString(R.string.hyphen),
+            context.getString(R.string.hyphen),
+            null
+          )
+        )
         val adapter =
           AnswerOptionDropDownArrayAdapter(context, R.layout.drop_down_list_item, answerOptionList)
-        val selectedAnswer =
-          questionnaireViewItem.answers.singleOrNull()?.value?.displayString(header.context)
+        val selectedAnswerIdentifier =
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DropDownViewHolderFactory.kt:87]: alternative without using `idenitiferString`
```suggestion
        val selectedAnswerIdentifier =
          questionnaireViewItem.answers.singleOrNull()?.value?.let { it.id ?: it.displayString(context)}
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DropDownViewHolderFactory.kt [hunk 5]
```diff
@@ -103,7 +111,9 @@ internal object DropDownViewHolderFactory :
             )
             val selectedAnswer =
               questionnaireViewItem.answerOption
-                .firstOrNull { it.value.displayString(context) == selectedItem?.answerOptionString }
+                .firstOrNull {
+                  it.value.identifierString(context) == selectedItem?.answerIdentifierString
+                }
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DropDownViewHolderFactory.kt:116]: can we just set selectedAnswer to this, I think it is doing the same thing is filtering for the id:
```suggestion
  val selectedAnswer =
              questionnaireViewItem.answers.singleOrNull()?.value
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreTypes.kt (hunk 1 above)
Comment (omarismail94) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreTypes.kt:101]: we are also only using this function in DropDownViewHolderFactory to replace `displayString`. Does this function also need to be used in other factories such as AutoComplete, which seem to be doing similar logic?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DropDownViewHolderFactory.kt (hunk 3 above)
Comment (LZRS) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DropDownViewHolderFactory.kt:71]: One obvious case would be when using an x-fhir-query as answerExpression for questionnaire items of type `reference`,  the resulting references in the answer options do not have their `Reference.id` populated and are null

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreTypes.kt (hunk 1 above)
Comment (LZRS) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreTypes.kt:116]: > do we need this function? check what I've done in the DropDownViewHolderFactory on how to get rid of it

For some types, in our case the `Reference` type, I felt that sometimes when the `Reference.id` was null, the `displayString` would not be the most unique identifier especially when using `choiceColumn` extension as compared to `Reference.reference`. The same for `Coding` type

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DropDownViewHolderFactory.kt (hunk 5 above)
Comment (LZRS) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DropDownViewHolderFactory.kt:116]: This fails initially when `questionnaireVIewItem.answers` is empty, essentially it never gets to select new

**Comment by LZRS** [comment]
@omarismail94 I've added the relevant description and addressed the comments

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreTypes.kt [hunk 6]
```diff
@@ -109,6 +109,18 @@ fun Type.getValueAsString(context: Context): String =
     else -> context.getString(R.string.not_answered)
   }
 
+/**
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreTypes.kt:112]: we can refactor this file quite a bit to remove the duplications that exist. I tried the below and it passed all the tests. If you think it works, can you use the suggested block below?

```suggestion
fun Type.displayString(context: Context): String =
  getDisplayString(this, context) ?: context.getString(R.string.not_answered)

/** Returns value as string depending on the [Type] of element. */
fun Type.getValueAsString(context: Context): String =
  getValueString(this) ?: context.getString(R.string.not_answered)

fun Type.identifierString(context: Context): String =
  id ?: (this as? Coding)?.code ?: (this as? Reference)?.reference ?: displayString(context)

private fun getDisplayString(type: Type, context: Context): String? =
  when (type) {
    is Coding -> type.displayElement.getLocalizedText() ?: type.display ?: type.code
    is StringType -> type.getLocalizedText() ?: type.asStringValue()
    is DateType -> type.localDate?.format()
    is DateTimeType -> "${type.localDate.format()} ${type.localTime.toLocalizedString(context)}"
    is Reference -> type.display ?: type.reference
    is Attachment -> type.url
    is BooleanType -> {
      when (type.value) {
        true -> context.getString(R.string.yes)
        false -> context.getString(R.string.no)
        null -> null
      }
    }
    is Quantity -> type.value.toString()
    else -> (type as? PrimitiveType<*>)?.valueAsString
  }

private fun getValueString(type: Type): String? =
  when (type) {
    is DateType,
    is DateTimeType,
    is StringType -> type.asStringValue()
    is Quantity -> type.value.toString()
    else -> (type as? PrimitiveType<*>)?.valueAsString
  }
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 7]
```diff
@@ -253,9 +256,17 @@ object ResourceMapper {
         // Set initial value for the questionnaire item. Questionnaire items should not have both
         // initial value and initial expression.
         questionnaireItem.initial =
-          mutableListOf(
-            Questionnaire.QuestionnaireItemInitialComponent().setValue(it.asExpectedType())
-          )
+          when (questionnaireItem.type) {
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:259]: how about we remove the handling of the questionnaire logic to asExpectedType to make this portion of the code only look like this:
```
         val value = it.asExpectedType(questionnaireItem.type)
        questionnaireItem.initial =
              mutableListOf(
                Questionnaire.QuestionnaireItemInitialComponent().setValue(value)
              )
```


and refactor asExpectedType to take in a parameter to represent the expected type:
```
private fun Base.asExpectedType(questionnaireItemType: Questionnaire.QuestionnaireItemType? = null): Type {
  return when {
    questionnaireItemType == Questionnaire.QuestionnaireItemType.REFERENCE -> asExpectedReferenceType()
    this is Enumeration<*> -> toCoding()
    this is IdType -> StringType(idPart)
    else -> this as Type
  }
}

private fun Base.asExpectedReferenceType(): Type {
  return when {
    this.isResource -> {
      this@asExpectedReferenceType as Resource
      Reference().apply {
        reference =
          "${this@asExpectedReferenceType.resourceType}/${this@asExpectedReferenceType.logicalId}"
      }
    }
    this is IdType ->
      Reference().apply {
        reference =
          "${this@asExpectedReferenceType.resourceType}/${this@asExpectedReferenceType.idPart}"
      }
    else -> throw FHIRException("Expression supplied does not evaluate to IdType.")
  }
}

```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/AutoCompleteViewHolderFactory.kt [hunk 8]
```diff
@@ -204,3 +218,9 @@ internal object AutoCompleteViewHolderFactory :
       }
     }
 }
+
+internal data class AutoCompleteViewAnswerOption(val answerId: String, val answerDisplay: String) {
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/AutoCompleteViewHolderFactory.kt:222]: nit: add Kdoc

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreTypes.kt [hunk 9]
```diff
@@ -59,54 +56,45 @@ fun Type.asStringValue(): String {
...
+
+/** Returns value as string depending on the [Type] of element. */
+fun Type.getValueAsString(context: Context): String =
+  getValueString(this) ?: context.getString(R.string.not_answered)
+
+/*
+ * Returns the unique identifier of a [Type]. Used to differentiate between item answer options that
+ * may have similar display strings
+ */
+fun Type.identifierString(context: Context): String =
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreTypes.kt:69]: slightly worried we're not dealing with identify of codes carefully enough. can we make this function much more rigorous please?

for example, comparing two coding just by code but not by system doesn't meet the brief of returning "unique identifier".

**Pull request merged**